### DISCUSSION TITLE
Paramiko for hermetic SSH connections

### DISCUSSION THREAD

**Discussion by mattyclarkson** [labgrid-project/labgrid#1527] [Ideas]
Currently the `SSHDriver` calls out to the `ssh` binary to perform SSH connections.

[paramiko] is a pure Python library for providing SSH connections.

We could use [paramiko] in the `SSHDriver` to remove the need for a host installed `ssh`. Another option would be a `ParamikoDriver` that performs the same as `SSHDriver` but with [paramiko].

`labgrid-client ssh` naturally forwards on arguments to the host `ssh` so would likely stay that way.

Would be of interest to project to investigate usage of `paramiko`?

[paramiko]: https://docs.paramiko.org/en/stable/

**Comment by Emantor** [comment]
We evaluated this in 2016 when labgrid was initially written. Paramiko AFAICS still has no support for the control socket infrastructure where connections are multiplexed over one initial connection. Background was that we wanted labgrid to reuse an existing open connection to the target if possible.

This was important to us then, but might not be as important now, unfortunately I don't know how much this feature is used currently.

**Reply by jluebbe** [reply]
I use this all the time, as it significantly improves performances, especially when using SSH's `ProxyJump` mechanism.

Also, we use SSH as a backend for rsync and for port forwarding, which wouldn't be easy if we were using Paramiko in the labgrid Python process.

**Reply by mattyclarkson** [reply]
Would a `ParamikoDriver` be something of interest? It wouldn't be as efficient as the `SSHDriver` but would work in environments without `ssh` installed (also assuming that you do not need `labgrid-client {ssh,rsync}` functionality).

**Reply by jluebbe** [reply]
Without `ssh`, large parts of labgrid cannot be used, so I'd see `ssh` as a dependency. As we wouldn't use a `ParamikoDriver` in our setup, it would needs someone to commit to maintain it.
